### PR TITLE
feat: support websocket transport

### PR DESCRIPTION
**Pull request by driftluo**
Add support for WebSocket, now web applications can communicate directly with each other.

### CODE REVIEW

**Comment by doitian** [comment]
Can this be added as an optional feature?

File: multiaddr/src/protocol.rs [hunk 1]
```diff
@@ -16,6 +16,8 @@ const IP6: u32 = 0x29;
 const P2P: u32 = 0x01a5;
 const TCP: u32 = 0x06;
 const TLS: u32 = 0x01c0;
+const WS: u32 = 477;
```
Comment (doitian) [inline multiaddr/src/protocol.rs:19]: Not consistent here. Please use hex literal.

File: src/transports/ws.rs [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+            Poll::Ready(None) | Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl Stream for WebsocketListener {
+    type Item = std::result::Result<(Multiaddr, WsStream), io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        match self.poll_pending(cx) {
```
Comment (doitian) [inline src/transports/ws.rs:220]: > Not a code change suggestion

`if let` is simpler here

File: src/transports/mod.rs [hunk 3]
```diff
@@ -48,22 +56,39 @@ impl Transport for MultiTransport {
...
+        match find_type(&address) {
+            TransportType::Tcp => match TcpTransport::new(self.timeout).listen(address) {
+                Ok(future) => Ok(MultiListenFuture::Tcp(future)),
+                Err(e) => Err(e),
+            },
+            TransportType::Ws => match WsTransport::new(self.timeout).listen(address) {
+                Ok(future) => Ok(MultiListenFuture::Ws(future)),
+                Err(e) => Err(e),
+            },
+            TransportType::Wss => Err(TransportErrorKind::NotSupported(address)),
```
[the remark this answers is not in the document]
Comment (driftluo) [reply src/transports/mod.rs:68]: > How about just removing these two transport types from the enum?

Explicit errors are better than strange behaviors, such as tls address recognized as tcp

File: src/transports/mod.rs [hunk 4]
```diff
@@ -48,22 +56,39 @@ impl Transport for MultiTransport {
...
+                Ok(future) => Ok(MultiDialFuture::Ws(future)),
+                Err(e) => Err(e),
+            },
+            TransportType::Wss => Err(TransportErrorKind::NotSupported(address)),
+            TransportType::TLS => Err(TransportErrorKind::NotSupported(address)),
         }
     }
 }
 
 pub enum MultiListenFuture {
```
[the remark this answers is not in the document]
Comment (driftluo) [reply src/transports/mod.rs:89]: How to be simple and abstract at the same time, that's the challenge.

I don't see a good solution at the moment. The good news is that here are the details of the library's internal implementation, which is sufficient for use at the moment.

**Review by doitian — APPROVED** [review]

File: src/transports/mod.rs [hunk 5]
```diff
@@ -92,18 +133,27 @@ impl Future for MultiDialFuture {
...
+            )
+            .poll(cx),
         }
     }
 }
 
 pub enum MultiStream {
     Tcp(TcpStream),
+    #[cfg(ws)]
+    Ws(Box<WsStream>),
```
Comment (quake) [inline src/transports/mod.rs:148]: unnecessary Box here?

Comment (driftluo) [reply src/transports/mod.rs:148]: clippy says it's too big here

File: src/transports/mod.rs [hunk 6]
```diff
@@ -92,18 +133,27 @@ impl Future for MultiDialFuture {
...
+                &mut inner.map(|res| res.map(|res| (res.0, MultiStream::Ws(Box::new(res.1))))),
+            )
+            .poll(cx),
         }
     }
 }
 
 pub enum MultiStream {
     Tcp(TcpStream),
+    #[cfg(ws)]
```
Comment (quake) [inline src/transports/mod.rs:147]: ws is a feature, i'm confused how this macro works here? should be:
```
#[cfg(feature = "ws")]
```

**Review by quake — CHANGES_REQUESTED** [review]

File: src/transports/ws.rs [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+        }
+
+        // calculate number of bytes that we can copy
+        let n = ::std::cmp::min(buf.len(), self.recv_buf.len());
+
+        // Copy data to the output buffer
+        buf[..n].copy_from_slice(self.recv_buf[..n].as_ref());
+
+        // drain n bytes of recv_buf
+        self.recv_buf = self.recv_buf.split_off(n);
```
Comment (quake) [inline src/transports/ws.rs:323]:
```suggestion
buf[..n].copy_from_slice(self.recv_buf.drain(..n).as_slice());
```

**Review by quake — APPROVED** [review]

File: src/transports/ws.rs (hunk 7 above)
Comment (driftluo) [reply src/transports/ws.rs:323]: I will use another PR to upgrade the toolchain and modify the related code

**Pull request merged**